data/core/commands/doc.lua

```diff
@@ -123,16 +124,24 @@ local commands = {
   ["doc:paste"] = function()
     local clipboard = system.get_clipboard()
     -- If the clipboard has changed since our last look, use that instead
-    if doc().cursor_clipboard["full"] ~= clipboard then
-      doc().cursor_clipboard = {}
-      doc().cursor_clipboard_whole_line = {}
+    if core.cursor_clipboard["full"] ~= clipboard then
+      core.cursor_clipboard = {}
+      core.cursor_clipboard_whole_line = {}
     end
+    local value, whole_line
     for idx, line1, col1, line2, col2 in doc():get_selections() do
-      local value = doc().cursor_clipboard[idx] or clipboard
-      if doc().cursor_clipboard_whole_line[idx] == true then
+      if #core.cursor_clipboard_whole_line == (#doc().selections/4) then
+        value = core.cursor_clipboard[idx]
+        whole_line = core.cursor_clipboard_whole_line[idx] == true
+      else
+        value = clipboard
+        whole_line = clipboard:find("\n") ~= nil
+      end
+      if whole_line then
         doc():insert(line1, 1, value:gsub("\r", ""))
-      elseif (core.lines_in_clipboard == clipboard) and (not doc().cursor_clipboard[idx]) then
-        doc():insert(line1, 1, clipboard:gsub("\r", ""))
+        if col1 == 1 then
+          doc():move_to_cursor(idx, #value)
+        end
```

Comment:
I don't like this thing. But without this, the cursor will either stay in place (if you do nothing), or go down (if you move it without checking). There might be something wrong with the insertion mechanism.
I suppose that the insertion occurs before the cursor(col1-1), but if the cursor is at the beginning of the line, then there is nowhere to subtract and the insertion occurs after.